Add a Usergreeting custome  module

kingfisher88 opened the pull request.

kingfisher88 committed:
Add a Usergreeting custome  module

lekob commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,32 @@
+ * GreetingTime is a greeting service for users
+ * dependent on the time of the day
+ */
+class greetingTime {
```
Your class is called GreetingTime, not greetingTime.

kingfisher88 committed:
Block without DI

rename usergreeting.service.yml to usergreeting.services.yml
rename a class in GreetingTime.php
make some changes in build() function in UsergreetingBlock.php

kingfisher88 committed:
Add username to block with dependency injection and change the conditions
to greet a user.

kingfisher88 committed:
code fixing with Code Beautifier

kingfisher88 committed:
Add missing short description in doc comment

kingfisher88 committed:
Correct the code with Code Sniffer and Code Beutifier.

kingfisher88 committed:
Add Dependensy Injection to UsergreetingBlock.php and GreetingTime.php

lekob commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,49 @@
+   * {@inheritdoc}
+   */
+  public function greetingMessage() {
+    $time_output = intval(date('H', $this->timeOutput->getRequestTime()));
```
uset type casting instead.
```suggestion
    $time_output = (int) date('H', $this->timeOutput->getRequestTime());
```

lekob replied:
And don't forget to factor in the user's timezone when calculating the greeting.

lekob commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,49 @@
+  /**
+   * The user account.
+   *
+   * @var \Drupal\usergreeting\Drupal\Component\Datetime\TimeInterface
```
This class does not exist

lekob commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,49 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function greetingMessage() {
```
missing type hint

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  /**
+   * {@inheritdoc}
```
This class doesn't extend/implement any other classes. Where does it inherit documentation from?

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Provides a greeting service for users dependent on the time of  day.
+ */
+class GreetingTime {
```
Is there any class that extends this class? If not, it's a good practice to make it `final`.

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,49 @@
+  protected $timeOutput;
+
+  /**
+   * Construct the Timeinterface object.
```
Do you construct the TimeInterface object? Or something else maybe? :)

gkurucz commented on web/modules/custom/usergreeting/src/Plugin/Block/UsergreetingBlock.php:
```diff
@@ -0,0 +1,81 @@
+  protected $account;
+
+  /**
+   * The user account.
```
This is the user account as well?

kingfisher88 committed:
Fix comments, type casting, type hint, and change date() to date_default_timezone_get()
function in GreetinTime.php
Fix comments in UsergreetingBlock.php
Change the font size in a Usergreeting block.

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,52 @@
+   *   The greeting message output.
+   */
+  final public function greetingMessage(string $greeting = NULL): greetingMessage {
+    $time_output = (int) date_default_timezone_get('H', $this->timeOutput->getRequestTime());
```
Unfortunately this will still not return the timezone of the current user. However I also noticed that the whole implementation could be better structured.

Instead of injecting the AccountInterface service to the Block plugin, why not do it here? Then, in the block you'd only need to get the message and render it, no need for additional actions.

Also, creating a string with whitespace at the end is not a good practice (line 41, 44 and 47). There are plenty of ways to join strings with a delimiter. (hint: https://www.php.net/manual/en/function.implode.php)

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,52 @@
+    $time_output = (int) date_default_timezone_get('H', $this->timeOutput->getRequestTime());
+
+    if ($time_output >= 5 && $time_output < 12) {
+      $greeting = 'Good Morning ';
```
Writing out strings as is not a good practice. What if this string needs to be translated? @lekob might be able to help more.

kingfisher88 committed:
Inject AccountInterface service to GreetinTime.php
Use StringTranslationTrait in GreetingTime.php
Use implode() at return for greetingMessage()
Make GreetinTime class final

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,67 @@
+ */
+final class GreetingTime {
+
+  use StringTranslationTrait;
```
Good direction! It won't be enough though unfortunately according to: https://api.drupal.org/api/drupal/core!lib!Drupal!Core!StringTranslation!StringTranslationTrait.php/trait/StringTranslationTrait/8.8.x

Most notably this line: 

> If the class is capable of injecting services from the container, it should inject the 'string_translation' service and assign it to $this->stringTranslation.

You're nearly there. :)

gkurucz commented on web/modules/custom/usergreeting/src/Plugin/Block/UsergreetingBlock.php:
```diff
@@ -0,0 +1,68 @@
+    $build = [];
+    $greetuser = $this->greeting->greetingMessage();
+
+    $build['#markup'] = '<p>' . $greetuser . '</p>';
```
Using `#markup` in render arrays is unsafe as it might not be filtered according to [the Markup object docs](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Markup.php/class/Markup/8.2.x). Let's say the username somehow contains XSS code which is then rendered on the page, resulting in the script executed. For your case it's preferred to use a different element for rendering, e.g.:

https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Element%21HtmlTag.php/class/HtmlTag/8.2.x

gkurucz commented on web/modules/custom/usergreeting/src/GreetingTime.php:
```diff
@@ -0,0 +1,67 @@
+   *   The greeting message output.
+   */
+  public function greetingMessage(): string {
+    $time_output = (int) date('G', $this->timeOutput->getRequestTime());
```
I checked the code, logged in with a user, changed my timezone to Vancouver and it's still Good evening. :) You still need to use the current user's timezone and format it so the right time is calculated.


This can probably help: 
https://api.drupal.org/api/drupal/core%21modules%21user%21src%21Entity%21User.php/function/User%3A%3AgetTimeZone/8.2.x
https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Datetime%21DateFormatter.php/function/DateFormatter%3A%3Aformat/8.2.x

gkurucz replied:
Berci told me you checked this together since then, it seems to be more of a caching issue and not a date issue then.

kingfisher88 committed:
Change the users timezone withoute cash rebuild.

kingfisher88 committed:
Merge branch 'master' of github.com:Pronovix/food into issue/8071

kingfisher88 committed:
Merge branch 'master' of github.com:Pronovix/food into issue/8071

kingfisher88 committed:
Add cache_context in GreetingTime.php
Add cache max-age in UsergreetingBlock.php

kingfisher88 committed:
Merge branch 'master' of github.com:Pronovix/food into issue/8071

kingfisher88 committed:
Correct the code according to PHPCS

kingfisher88 committed:
Uncomment $build ['#cache']['max-age'] = 0; in UsergreetingBlock.php

kingfisher88 committed:
Uncomment $build ['#cache']['max-age'] = 0; in UsergreetingBlock.php
and remove uneccessery space
